File: quant1x/data/meta/timestamp.py
# -*- coding: UTF-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import time
from datetime import datetime, timedelta
from typing import Optional, Tuple, Union
# Pre-market time constants
from .tradinghours import PRE_MARKET_HOUR, PRE_MARKET_MINUTE, PRE_MARKET_SECOND

# Constants
SECONDS_PER_MINUTE = 60
SECONDS_PER_HOUR = 60 * 60
SECONDS_PER_DAY = 24 * 60 * 60
MILLISECONDS_PER_SECOND = 1000
MILLISECONDS_PER_MINUTE = 60 * 1000
MILLISECONDS_PER_HOUR = 60 * 60 * 1000
MILLISECONDS_PER_DAY = 24 * 60 * 60 * 1000


class Timestamp:
    """
    Timestamp class compatible with C++/Rust/Go implementations.
    Stores time as LOCAL milliseconds (not UTC!), similar to C++ implementation.
    Strings are parsed as local time. When converting to/from UTC,
    use the conversion functions.
    """
    def __init__(self, ms: Union[int, float] = 0):
        self.ms = int(ms)

    @staticmethod
    def now() -> 'Timestamp':
        """Get current timestamp (LOCAL milliseconds)"""
        # Get current local datetime, then convert to local milliseconds
        dt_local = datetime.now()
        return Timestamp.from_datetime(dt_local)

    @staticmethod
    def zero() -> 'Timestamp':
        return Timestamp(0)

    def value(self) -> int:
        return self.ms

    def to_datetime(self) -> datetime:
        """Convert local milliseconds to local datetime (naive)"""
        seconds = self.ms / 1000.0
        # epoch = 1970-01-01 00:00:00 LOCAL time
        epoch = datetime(1970, 1, 1)
        return epoch + timedelta(seconds=seconds)

    @staticmethod
    def from_datetime(dt: datetime) -> 'Timestamp':
        """Convert local datetime to local milliseconds"""
        # dt is assumed to be local time (naive)
        # We want to store as LOCAL milliseconds, not UTC
        # epoch = 1970-01-01 00:00:00 LOCAL time
        epoch = datetime(1970, 1, 1)
        local_seconds = (dt - epoch).total_seconds()
        return Timestamp(local_seconds * 1000)

    def start_of_day(self) -> 'Timestamp':
        """Get the timestamp of 00:00:00 on the same day (Local time)"""
        return Timestamp(self.ms - (self.ms % MILLISECONDS_PER_DAY))

    @staticmethod
    def midnight() -> 'Timestamp':
        """Get the timestamp of today's 00:00:00 (Local time)"""
        ts = Timestamp.now()
        return Timestamp(ts.ms - (ts.ms % MILLISECONDS_PER_DAY))

    def today(self, hour: int = 0, minute: int = 0, second: int = 0, millisecond: int = 0) -> 'Timestamp':
        """Get timestamp for specific time on the same day (Local time)"""
        ts = self.start_of_day().value()
        ts += hour * MILLISECONDS_PER_HOUR
        ts += minute * MILLISECONDS_PER_MINUTE
        ts += second * MILLISECONDS_PER_SECOND
        ts += millisecond
        return Timestamp(ts)

    def since(self, hour: int = 0, minute: int = 0, second: int = 0, millisecond: int = 0) -> 'Timestamp':
        """Alias for today()"""
        return self.today(hour, minute, second, millisecond)

    def offset(self, hour: int = 0, minute: int = 0, second: int = 0, millisecond: int = 0) -> 'Timestamp':
        """Add offset to current timestamp (in milliseconds)"""
        ts = self.value()
        ts += hour * MILLISECONDS_PER_HOUR
        ts += minute * MILLISECONDS_PER_MINUTE
        ts += second * MILLISECONDS_PER_SECOND
        ts += millisecond
        return Timestamp(ts)

    @staticmethod
    def pre_market_time(year: int, month: int, day: int) -> 'Timestamp':
        """Construct pre-market timestamp (09:00:00) for specific date"""
        dt = datetime(year, month, day, PRE_MARKET_HOUR, PRE_MARKET_MINUTE, PRE_MARKET_SECOND)
        return Timestamp.from_datetime(dt)

    def get_pre_market_time(self) -> 'Timestamp':
        """Get pre-market timestamp for the same day"""
        return self.today(PRE_MARKET_HOUR, PRE_MARKET_MINUTE, PRE_MARKET_SECOND, 0)

    def floor(self) -> 'Timestamp':
        """Round down to nearest minute (00 seconds, 000 ms)"""
        ts = self.value()
        ts -= (ts % MILLISECONDS_PER_MINUTE)
        return Timestamp(ts)

    def ceil(self) -> 'Timestamp':
        """Round up to end of minute (59 seconds, 999 ms)"""
        ts = self.value()
        ts = ts - (ts % MILLISECONDS_PER_MINUTE) + (MILLISECONDS_PER_MINUTE - 1)
        return Timestamp(ts)

    def extract(self) -> Tuple[int, int, int]:
        """Return (year, month, day) in local time"""
        dt = self.to_datetime()
        return dt.year, dt.month, dt.day

    def to_string(self, layout: str = "%Y-%m-%d %H:%M:%S") -> str:
        """Convert to string in local time"""
        dt = self.to_datetime()
        return dt.strftime(layout)

    def only_date(self) -> str:
        return self.to_string("%Y-%m-%d")

    def cache_date(self) -> str:
        return self.to_string("%Y%m%d")

    def only_time(self) -> str:
        """Return time only, truncated to seconds"""
        dt = self.to_datetime()
        return dt.strftime("%H:%M:%S")
    
    def yyyymmdd(self) -> int:
        """返回日期部分的整数表示，格式为 YYYYMMDD"""
        dt = self.to_datetime()
        return dt.year * 10000 + dt.month * 100 + dt.day

    def is_empty(self) -> bool:
        return self.ms == 0

    def is_same_date(self, other: 'Timestamp') -> bool:
        """Check if two timestamps are on the same day (local time)"""
        day1 = self.ms // MILLISECONDS_PER_DAY
        day2 = other.ms // MILLISECONDS_PER_DAY
        return day1 == day2

    @staticmethod
    def parse(time_str: str) -> 'Timestamp':
        """Parse time string as local time"""
        formats = [
            "%Y-%m-%d %H:%M:%S.%f",
            "%Y-%m-%d %H:%M:%S",
            "%Y-%m-%d",
            "%Y%m%d",
            "%Y/%m/%d %H:%M:%S",
            "%Y/%m/%d",
            "%m/%d/%Y %H:%M:%S",
        ]
        for fmt in formats:
            try:
                dt = datetime.strptime(time_str, fmt)
                return Timestamp.from_datetime(dt)
            except ValueError:
                continue
        raise ValueError(f"Unable to parse timestamp: {time_str}")

    @staticmethod
    def parse_time(time_str: str) -> 'Timestamp':
        """Parse time string, assuming today's date if date is missing"""
        # Try full formats first
        try:
            return Timestamp.parse(time_str)
        except ValueError:
            pass

        # Try time-only formats
        time_formats = [
            "%H:%M:%S.%f",
            "%H:%M:%S",
            "%H:%M",
            "%H%M%S",
            "%H%M",
        ]

        now = datetime.now()
        for fmt in time_formats:
            try:
                t = datetime.strptime(time_str, fmt).time()
                dt = datetime.combine(now.date(), t)
                return Timestamp.from_datetime(dt)
            except ValueError:
                continue
        raise ValueError(f"Unable to parse time string: {time_str}")

    # Comparison operators
    def __eq__(self, other):
        if isinstance(other, Timestamp):
            return self.ms == other.ms
        return False

    def __lt__(self, other):
        if isinstance(other, Timestamp):
            return self.ms < other.ms
        return NotImplemented

    def __le__(self, other):
        if isinstance(other, Timestamp):
            return self.ms <= other.ms
        return NotImplemented

    def __gt__(self, other):
        if isinstance(other, Timestamp):
            return self.ms > other.ms
        return NotImplemented

    def __ge__(self, other):
        if isinstance(other, Timestamp):
            return self.ms >= other.ms
        return NotImplemented

    def __repr__(self):
        return f"Timestamp({self.ms}, '{self.to_string()}')"

    def __str__(self):
        return self.to_string()


if __name__ == "__main__":
    ts = Timestamp.parse("1970-01-01")
    print(ts)

    ts = Timestamp.parse("1900-01-01")
    print(ts)

    ts = Timestamp.parse("2024-01-01 09:30:00")
    print(ts)

    ts = Timestamp.now()
    print(f"Now: {ts}")
    
    ts = Timestamp.zero()
    print(f"Zero: {ts}")

File: quant1x/data/meta/timestamp.cpp
#include <quant1x/data/meta/timestamp.h>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <sstream>
#include <iomanip>

namespace meta {

Timestamp::Timestamp(const std::string& s) {
    if (s.empty()) {
        ms = 0;
        return;
    }
    *this = parse(s);
}

Timestamp Timestamp::now() {
    auto now_tp = std::chrono::system_clock::now();
    auto ms_tp = std::chrono::duration_cast<std::chrono::milliseconds>(now_tp.time_since_epoch());
    // 转换为本地时间
    auto now_c = std::chrono::system_clock::to_time_t(now_tp);
    std::tm local_tm{};
#ifdef _WIN32
    localtime_s(&local_tm, &now_c);
#else
    localtime_r(&now_c, &local_tm);
#endif
    // 计算本地时区偏移
    std::time_t utc_mktime = std::mktime(&local_tm);
    auto utc_offset = static_cast<int64_t>(std::difftime(utc_mktime, now_c)) * MILLISECONDS_PER_SECOND;
    return Timestamp(ms_tp.count() + utc_offset);
}

Timestamp Timestamp::parse(const std::string& s) {
    if (s.empty()) return zero();
    std::tm tm{};
    [[maybe_unused]] int ms_part = 0;
    // 尝试多种格式
    const char* formats[] = {
        "%Y-%m-%d %H:%M:%S", "%Y-%m-%d", "%Y%m%d", "%Y/%m/%d %H:%M:%S",
        "%Y/%m/%d", "%Y%m%d %H%M%S", "%Y-%m-%dT%H:%M:%S"
    };
    for (auto fmt : formats) {
        std::istringstream ss(s);
        ss >> std::get_time(&tm, fmt);
        if (!ss.fail()) {
            std::time_t t = std::mktime(&tm);
            auto tp = std::chrono::system_clock::from_time_t(t);
            auto ms_tp = std::chrono::duration_cast<std::chrono::milliseconds>(tp.time_since_epoch());
            return Timestamp(ms_tp.count());
        }
    }
    // 最后尝试只有日期的紧凑格式
    if (s.length() == 8 && std::all_of(s.begin(), s.end(), ::isdigit)) {
        int year = std::stoi(s.substr(0, 4));
        int month = std::stoi(s.substr(4, 2));
        int day = std::stoi(s.substr(6, 2));
        return from_date(year, month, day);
    }
    return zero();
}

Timestamp Timestamp::parse_time(const std::string& s) {
    if (s.empty()) return zero();
    std::tm tm{};
    [[maybe_unused]] int ms_part = 0;
    // 先尝试纯时间格式
    const char* time_formats[] = {"%H:%M:%S", "%H:%M", "%H%M%S"};
    for (auto fmt : time_formats) {
        std::istringstream ss(s);
        ss >> std::get_time(&tm, fmt);
        if (!ss.fail()) {
            auto now_tp = std::chrono::system_clock::now();
            auto now_c = std::chrono::system_clock::to_time_t(now_tp);
            std::tm local_tm{};
#ifdef _WIN32
            localtime_s(&local_tm, &now_c);
#else
            localtime_r(&now_c, &local_tm);
#endif
            local_tm.tm_hour = tm.tm_hour;
            local_tm.tm_min = tm.tm_min;
            local_tm.tm_sec = tm.tm_sec;
            std::time_t t = std::mktime(&local_tm);
            auto tp = std::chrono::system_clock::from_time_t(t);
            auto ms_tp = std::chrono::duration_cast<std::chrono::milliseconds>(tp.time_since_epoch());
            return Timestamp(ms_tp.count());
        }
    }
    // 回退到完整日期时间解析
    return parse(s);
}

Timestamp Timestamp::from_date(int year, int month, int day,
                                int hour, int minute, int second, int millisecond) {
    std::tm tm{};
    tm.tm_year = year - 1900;
    tm.tm_mon = month - 1;
    tm.tm_mday = day;
    tm.tm_hour = hour;
    tm.tm_min = minute;
    tm.tm_sec = second;
    std::time_t t = std::mktime(&tm);
    auto tp = std::chrono::system_clock::from_time_t(t);
    auto ms_tp = std::chrono::duration_cast<std::chrono::milliseconds>(tp.time_since_epoch());
    return Timestamp(ms_tp.count() + millisecond);
}

Timestamp Timestamp::pre_market_time() const {
    int year, month, day;
    extract(year, month, day);
    return from_date(year, month, day, PRE_MARKET_HOUR, PRE_MARKET_MINUTE, PRE_MARKET_SECOND, 0);
}

Timestamp Timestamp::pre_market_time(int year, int month, int day) {
    return from_date(year, month, day, PRE_MARKET_HOUR, PRE_MARKET_MINUTE, PRE_MARKET_SECOND, 0);
}

std::string Timestamp::only_date() const {
    return to_string("%Y-%m-%d");
}

std::string Timestamp::only_time() const {
    return to_string("%H:%M:%S");
}

std::string Timestamp::cache_date() const {
    return to_string("%Y%m%d");
}

int64_t Timestamp::yyyymmdd_int() const {
    int year, month, day;
    extract(year, month, day);
    return static_cast<int64_t>(year) * 10000 + month * 100 + day;
}

Timestamp Timestamp::from_yyyymmdd_int(int64_t date) {
    int year = static_cast<int>(date / 10000);
    int month = static_cast<int>((date % 10000) / 100);
    int day = static_cast<int>(date % 100);
    return from_date(year, month, day);
}

std::string Timestamp::to_string(const char* layout) const {
    std::time_t t = static_cast<std::time_t>(ms / MILLISECONDS_PER_SECOND);
    std::tm local_tm{};
#ifdef _WIN32
    localtime_s(&local_tm, &t);
#else
    localtime_r(&t, &local_tm);
#endif
    std::ostringstream oss;
    oss << std::put_time(&local_tm, layout);
    return oss.str();
}

bool Timestamp::is_same_date(const Timestamp& other) const {
    int y1, m1, d1, y2, m2, d2;
    extract(y1, m1, d1);
    other.extract(y2, m2, d2);
    return y1 == y2 && m1 == m2 && d1 == d2;
}

Timestamp Timestamp::offset(int hour, int minute, int second, int millisecond) const {
    int64_t offset_ms = static_cast<int64_t>(hour) * MILLISECONDS_PER_HOUR +
                        static_cast<int64_t>(minute) * MILLISECONDS_PER_MINUTE +
                        static_cast<int64_t>(second) * MILLISECONDS_PER_SECOND +
                        millisecond;
    return Timestamp(ms + offset_ms);
}

Timestamp Timestamp::start_of_day() const {
    return today_at(0, 0, 0, 0);
}

Timestamp Timestamp::today_at(int hour, int minute, int second, int millisecond) const {
    int year, month, day;
    extract(year, month, day);
    return from_date(year, month, day, hour, minute, second, millisecond);
}

void Timestamp::extract(int& year, int& month, int& day) const {
    std::time_t t = static_cast<std::time_t>(ms / MILLISECONDS_PER_SECOND);
    std::tm local_tm{};
#ifdef _WIN32
    localtime_s(&local_tm, &t);
#else
    localtime_r(&t, &local_tm);
#endif
    year = local_tm.tm_year + 1900;
    month = local_tm.tm_mon + 1;
    day = local_tm.tm_mday;
}

void Timestamp::extract_time(int& hour, int& minute, int& second, int& millisecond) const {
    std::time_t t = static_cast<std::time_t>(ms / MILLISECONDS_PER_SECOND);
    std::tm local_tm{};
#ifdef _WIN32
    localtime_s(&local_tm, &t);
#else
    localtime_r(&t, &local_tm);
#endif
    hour = local_tm.tm_hour;
    minute = local_tm.tm_min;
    second = local_tm.tm_sec;
    millisecond = static_cast<int>(ms % MILLISECONDS_PER_SECOND);
}

} // namespace meta
